### ISSUE TITLE
Bad interaction between drreg and ISB mangling on AArch64

### ISSUE THREAD

**Issue by egrimley** [DynamoRIO/dynamorio#2415]
There seems to be a bad interaction between drreg (used by bbcount.c in this example) and the mangling of ISB (see mangle_icache_op). It looks as though drreg thinks that X0 and X1 do not need to be restored before the ISB, but they do, because of the way ISB is mangled.

What's the right way of solving this?

The original basic block is just:
```
  0xa60894:  dsb     ish
  0xa60898:  isb
  0xa6089c:  ret
```
The mangled and instrumented version in the fragment cache is:
```
0x4be0700c:  str     x0, [x28,#272]        <-- drreg saves X0
0x4be07010:  str     x1, [x28,#280]        <-- drreg saves X1
0x4be07014:  mov     x0, #0x1490           <-- start of code inserted by drx_insert_counter_update
0x4be07018:  movk    x0, #0x7201, lsl #16
0x4be0701c:  ldr     x1, [x0]
0x4be07020:  add     x1, x1, #0x1
0x4be07024:  str     x1, [x0]              <-- end of instrumentation code; x0 and x1 are not restored!
0x4be07028:  dsb     ish
0x4be0702c:  isb                           <-- start of mangled ISB
0x4be07030:  str     x0, [x28]             <-- corrupted X0 is saved; it is later retrieved by icache_op_isb_asm
0x4be07034:  mov     x0, #0x2b20
0x4be07038:  movk    x0, #0x559c, lsl #16
0x4be0703c:  movk    x0, #0x55, lsl #32
0x4be07040:  ldr     w0, [x0]
0x4be07044:  cbz     w0, 0x4be07064        <-- if we took this branch there would be no problem
0x4be07048:  stp     x1, x2, [x28,#8]      <-- corrupted X1 is saved; it is later retrieved by icache_op_isb_asm
0x4be0704c:  mov     x2, #0x2698
0x4be07050:  movk    x2, #0x5587, lsl #16
0x4be07054:  movk    x2, #0x55, lsl #32
0x4be07058:  mov     x1, #0x0
0x4be0705c:  mov     x0, x28
0x4be07060:  br      x2                    <-- branch to icache_op_isb_asm
0x4be07064:  ldr     x0, [x28]             <-- end of mangled ISB
0x4be07068:  ldr     x0, [x28,#272]        <-- X0 is restored by drreg, but it's too late!
0x4be0706c:  ldr     x1, [x28,#280]        <-- X1 is restored by drreg, but it's too late!
0x4be07070:  str     x2, [x28,#16]         <-- start of mangled RET
0x4be07074:  mov     x2, x30
0x4be07078:  b       0x4be0707c
0x4be0707c:  stp     x0, x1, [x28]
0x4be07080:  mov     x0, #0xd230
0x4be07084:  movk    x0, #0x4be0, lsl #16
0x4be07088:  movk    x0, #0x0, lsl #32
0x4be0708c:  movk    x0, #0x0, lsl #48
0x4be07090:  ldr     x1, [x28,#120]
0x4be07094:  br      x1
```

**Comment by derekbruening** [comment]
At first this seemed like something unique to isb sneakily exiting the cache behind the client's back, but thinking further there are cases on x86 where this happens (though through an exit cti: are we sure this non-exit-cti is safe?) and I don't think they are correctly handled wrt instrumentation.

This is not specific to drreg: instrumentation in general can have state across app instructions.

For mangling that adds an exit out of the cache, such as this OP_isb, or x86 selfmod, or inlined syscall skipping, the simplest thing would be for the instruction to already potentially fault and then for DR to invoke the restore state event back in dispatch.  It would use the mcontext at that point which presumably would cover post-write and not just pre-write for selfmod.

For instrs that don't normally fault, the client can reasonably refuse to translate (for synchall where a retry is possible).  I guess we'd have to just not support such a situation, which would require listing all places where we need to translate in the docs.

**Comment by egrimley** [comment]
It would be nice to have a temporary work-around for this problem, either just for the AArch64 ISB case, or more generally. It doesn't even have to be something committed to master: a small patch applied locally might let us use DynamoRIO for instrumenting a self-modifying program or at least let us uncover any further problems. Any ideas for a neat way of doing that? A simple way of telling the code that ISB is really dangerous, perhaps?

**Comment by egrimley** [comment]
The problem can be demonstrated on AArch64 by attempting almost any kind of instrumentation on a self-modifying program, for example:
```
$ bin64/drrun -c api/bin/libbbcount.so -- suite/tests/bin/security-common.codemod
starting up
Segmentation fault (core dumped)
```

**Comment by derekbruening** [comment]
> It would be nice to have a temporary work-around for this problem

?  The permanent solution is simple and straightforward.  I don't see how any work-around is going to be as simple or worth the effort rather than putting that effort into directly implementing the actual solution.

**Comment by egrimley** [comment]
So what is the permanent solution?

**Comment by derekbruening** [comment]
https://github.com/DynamoRIO/dynamorio/issues/2415#issuecomment-299483713

**Comment by egrimley** [comment]
Yes, but I don't see any relevant mentions of "fault" in instr.h or drreg.c so I'm not sure what kind of change you're suggesting. Is there a synonym for "fault" I should grep for?

**Comment by derekbruening** [comment]
The restore state event is a basic part of DR's API:

http://dynamorio.org/docs/API_BT.html#sec_translation

https://github.com/DynamoRIO/dynamorio/blob/master/core/lib/instrument_api.h#L292

https://github.com/DynamoRIO/dynamorio/blob/master/core/lib/instrument_api.h#L573

https://github.com/DynamoRIO/dynamorio/blob/master/ext/drreg/drreg.c#L1360
